xrmocap/model/mvp/projattn.py

```diff
@@ -0,0 +1,281 @@
+            query.unsqueeze(2)).view(n_views, Len_q, self.n_heads,
+                                     feat_lvls * self.n_points)
+        attention_weights = F.softmax(attention_weights,
+                                      -1).view(n_views, Len_q, self.n_heads,
+                                               feat_lvls, self.n_points)
+
+        if reference_points.shape[-1] == 2:
+            offset_normalizer = torch.stack(
+                [input_spatial_shapes[..., 1], input_spatial_shapes[..., 0]],
+                -1)
+            sampling_locations = reference_points[:, :, None, :, None, :] \
+                + sampling_offsets \
+                / offset_normalizer[None, None, None, :, None, :]
+        elif reference_points.shape[-1] == 4:
+            sampling_locations = reference_points[:, :, None, :, None, :2] \
+                + sampling_offsets / self.n_points \
+                * reference_points[:, :, None, :, None, 2:] * 0.5
+        else:
+            raise ValueError('Last dim of reference_points must be 2 or 4, \
+                but get {} instead.'.format(reference_points.shape[-1]))
+        output = DeformFunction.apply(value, input_spatial_shapes,
+                                      input_level_start_index,
+                                      sampling_locations, attention_weights,
+                                      self.im2col_step)
+        output = self.output_proj(output)
+        return output
+
+
+class DeformFunction(Function):
+
+    @staticmethod
+    def forward(ctx, value, value_spatial_shapes, value_level_start_index,
+                sampling_locations, attention_weights, im2col_step):
+        ctx.im2col_step = im2col_step
+        output = DF.deform_forward(value, value_spatial_shapes,
+                                   value_level_start_index, sampling_locations,
+                                   attention_weights, ctx.im2col_step)
+        ctx.save_for_backward(value, value_spatial_shapes,
+                              value_level_start_index, sampling_locations,
+                              attention_weights)
+        return output
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_output):
+        (value, value_spatial_shapes, value_level_start_index,
+         sampling_locations, attention_weights) = ctx.saved_tensors
+
+        grad_value, grad_sampling_loc, grad_attn_weight = \
+            DF.deform_backward(
+                value, value_spatial_shapes, value_level_start_index,
+                sampling_locations, attention_weights,
+                grad_output, ctx.im2col_step)
+
+        return \
+            grad_value, \
+            None, None, \
+            grad_sampling_loc, \
+            grad_attn_weight, None
+
+
+def _is_power_of_2(n):
+    if (not isinstance(n, int)) or (n < 0):
+        raise ValueError('invalid input '
+                         'for _is_power_of_2: '
+                         '{} (type: {})'.format(n, type(n)))
```

Comment:
f-string?

Reply:
Fixed in [d5025b7](https://github.com/openxrlab/xrmocap/pull/21/commits/d5025b700c504f02e48cdaae421d19a4d3d2c8d4).